### ISSUE TITLE
Cross entryp learning

### ISSUE THREAD

**Issue by mikerabat** [joaopauloschuler/neural-api#135]
I hope I'm not too annoying but you guys are the experts in that area so I hope I can discuss another neat feature with you...

While browsing through the "Neural Networks for Pattern Recognition" from C. M. Bishop I recognized that there are more than
the standard learning error propagation method with mean squared error but rather there is one called
Cross Entropy loss function... There are a few sources that claim that this error/loss function would indeed allow faster learning
progress....

What do you think? Would that be a viable feature for the library?

**Comment by joaopauloschuler** [comment]
The forward pass of categorical cross-entropy is implemented via `TNNetSoftMax`. There is an example at:  
https://github.com/joaopauloschuler/neural-api/blob/master/examples/SimpleImageClassifier/SimpleImageClassifier.lpr

Regarding the backpropagation, I have changed my mind about the best approach to it a number of times. In some APIs, the derivative on the last softmax layer is just not calculated and the errors are passed assuming derivative 1. You can get this behaviour via the parameter `{SkipBackpropDerivative=}1`. In most of cases, we get faster convergence with `{SkipBackpropDerivative=}1`.

You can use cross-entropy right now with `TNNetSoftMax` as per example.

**Comment by joaopauloschuler** [comment]
@mikerabat , you are certainly not annoying. Glad to help.

**Comment by mikerabat** [comment]
Thank you for the clarification. I guess that my misunderstanding was that I thought the cross-entropy type of error propagation would also be applied in the inner layers... I actually always struggled to wrap my head around that...
